recipes/psipred/perlwrapper.patch

```diff
@@ -0,0 +1,179 @@
++
++my $inputfile = undef;
++my $dbname = undef;
++my $ncbidir = "";
++my $mtx = 0;
++
++&get_arguments();
++
++# Generate a "unique" temporary filename root
++my $hostid = qx(uname -n); chomp $hostid;
++my $tmproot = ($ENV{TMPDIR} ? exists $ENV{TMPDIR} : '').'psitmp'.$$.$hostid;
++qx(mkdir -p $tmproot && cp -f $inputfile $tmproot/input); #($? != 0)
++
++if ($mtx != 1) {
++	blast();
++} else {
++	print "Re-using pre-computed PSI-BLAST with from file $inputfile ...\n";
++	qx(cp $inputfile $tmproot/psiblast.mtx);
++}
++my $resfiles = psipred();
++cleanup($resfiles);
++
++sub cleanup {
++	my ($resfiles) = @_;
++	
++	print "Cleaning up ...\n";
++	my @filesToDelete = ('input', 'psiblast.chk', 'psiblast.blast', 'psiblast.pn', 'psiblast.sn', 'psiblast.mn', 'psiblast.aux', 'pass1.ss');
++	push @filesToDelete, 'psiblast.mtx' if ($mtx != 1);
++	foreach my $file (@filesToDelete) {
++		qx(rm -f $tmproot/$file);
++	}
++	
++	my $pwd = qx(pwd); chomp $pwd; $pwd .= '/';
++	print "Final output files: ";
++	foreach my $file (@{$resfiles}) {
++		$file =~ s/^$pwd//;
++		print $file." ";
++	}
++	print "Finished.\n";
++}
++
++sub psipred {
++	print "Pass1 ...\n";
++
++	my $filename = $tmproot.'/'.basename($inputfile);
++	my $cmd = $execdir.'psipred '.$tmproot.'/psiblast.mtx '.$datadir.'weights.dat '.$datadir.'weights.dat2 '.$datadir.'/weights.dat3 > '.$tmproot.'/pass1.ss';
++	qx($cmd);
++	my $status = $?;
++	if ($status != 0) {
++		print "FATAL: Error whilst running psipred - script terminated!\n";
++		exit $status;
++	}
++
++	print "Pass2 ...\n";
++	$cmd = $execdir.'psipass2 '.$datadir.'weights_p2.dat 1 1.0 1.0 '.File::Spec->rel2abs($filename.'.ss2').' '.$tmproot.'/pass1.ss > '.File::Spec->rel2abs($filename.'.horiz');
++	qx($cmd);
++	$status = $?;
++	if ($status != 0) {
++		print "FATAL: Error whilst running psipass2 - script terminated!\n";
++		exit $status;
++	}
++	
++	return [File::Spec->rel2abs($filename.'.ss2'), File::Spec->rel2abs($filename.'.horiz')]
++}
++
++sub blast {
++	print "Running PSI-BLAST with sequence $inputfile ...\n";
++		
++	my $blastdatadir = "";
++	my $cmd = $ncbidir.'blastpgp -b 0 -v 5000 -j 3 -h 0.001 -d '.$dbname.' -i '.$tmproot.'/input -C '.$tmproot.'/psiblast.chk &> '.$tmproot.'/psiblast.blast';
++	qx($cmd);
++	my $status = $?;
++	if ($status != 0) {
++		system("tail ${tmproot}/psiblast.blast");
++		print "FATAL: Error whilst running blastpgp - script terminated!\n";
++		exit $status
++	}
++
++	print "Predicting secondary structure...\n";
++	qx(echo 'psiblast.chk' > $tmproot/psiblast.pn);
++	qx(echo 'input' > $tmproot/psiblast.sn);
++	
++	$cmd = $ncbidir.'makemat -P '.$tmproot.'/psiblast';
++	qx($cmd);
++	$status = $?;
++	if ($status != 0) {
++		print "FATAL: Error whilst running makemat - script terminated!\n";
++		exit $status;
++	}
++}
++
++sub get_arguments {
++	my $result = GetOptions (
++				"d=s" => \$dbname,
++				"n=s" => \$ncbidir,
++				"mtx=i" => \$mtx,
++		        "h"  => sub {&usage;});
++
++	&usage if (!$ARGV[0]);
```

Comment:
Be careful here. Unlike Python, Perl's `ARGV` starts with 0. That is, in command `perl script.pl infile outfile`. `$ARGV[0]` is `infile`. In your code, if the user doesn't enter any argument, this will return an undefined error.
Maybe you can do `&usage unless @ARGV;`